### ISSUE TITLE
Add a select command filter for previous selection

### ISSUE THREAD

**Issue by salinecitrine** [beyond-all-reason/spring#1205]
I was trying to answer a [question](https://discord.com/channels/549281623154229250/858397131148623892/1196162701680967730) about selecting the opposite of half-of-selected-units (the half selector being Ctrl+q in grid keys, or `select PrevSelection++_ClearSelection_SelectPart_50+`), and came up with what I thought was a solution:

`select AllMap+_InGroup_1_Not_InPrevSel+_ClearSelection_SelectAll+`

I imagined that this would select all units in the given control group (1 in this case) that are not currently selected.

But this doesn't work, because `InPrevSel` is actually this:

> Keep only units of the same type (unitDefID) as any unit in the selection before this select command was run.

which does not really match the name of the filter. It seems unlikely we'd change `InPrevSel` due to backwards compatibility concerns, but I think adding a filter for previously selected units would be helpful.

**Comment by salinecitrine** [comment]
Allowing multiple SOURCE sections (so you could do something like `select AllMap_Not_PrevSelection+_InGroup_1+_ClearSelection_SelectAll+`) would also address this.

**Comment by sprunk** [comment]
I think this is the point where selections should be done in Lua, because "previous selection" is often logically not what it is technically. Consider some cases:
* you have a tank selected. You select 2 bots. A moment later the game automatically deselects one because it got captured/stunned/whatever. Current selection is now 1 bot and previous selection is 2 bots (not the tank).
* you have a tank selected. You try to select a bot and a solar collector. This nominally selects these 2 units, but there is a filter that excludes the solar and implements it the sensible way by letting the engine perform the full selection and then deselecting the solar. Current selection is now the bot, and previous selection is bot and solar (not the tank).
* you have a tank selected. You drag a box over 5 bots. But you are playing BAR which has this stupid mechanic where units are selected 1-by-1 as you drag, so the previous selection is 4 bots (not the tank).

**Comment by salinecitrine** [comment]
My understanding was that "previous selection" in this context means the selected units as the `select` command is begun, and that no other selection mechanics would be active during the processing of the `select`. Like this:

From https://beyond-all-reason.github.io/spring/articles/select-command/
> PrevSelection: All units present in previous selection; that is, the one active before you hit the selection key.

I can imagine Lua selections running into the issues you mention, but I'm not seeing how the `select` command would.

**Comment by sprunk** [comment]
Yes, but "the selected units as the select command is begun" is not what the request you linked was about. It was about the units that were selected at some earlier point in time, and are the last **de**selected ones:

![image](https://github.com/beyond-all-reason/spring/assets/2573076/a14f3105-2ca0-4048-b28c-67da6112a48f)

But the set of that units can change without player input. To use the "select the other half of units" motivating example:
* T=0. You have units ABCD selected. You press the "select half" hotkey and this leaves AB. Now there is a set of units ABCD which I would normally call "the previous selection" but since existing filters use it for the current selection I am going to call it the "earlier" selection.
* T=1. B gets deselected somehow in one of the ways I mentioned above, which makes the current (or, as existing filters call it, previous) selection A, and the earlier selection AB. This can happen in ways that players might not necessarily consider deselection so the player still expects that the earlier selection is ABCD.
* T=2. Player wants to select the other half of ABCD, so uses the newly added "select earlier selection + NOT in previous (current) selection" filter, logically expecting that this would select CD, which it doesn't.

**Comment by salinecitrine** [comment]
Ok, I think I understand what you're saying. I agree that "previous selection" is too vague a term. In the context of this issue, only the `PrevSelection` style meaning (the selection as the `select` command is issued) is what I'm referring to, and the "earlier selection", while definitely relevant to selections and the discord conversation, is not a requirement for the hypothetical new `select` filter.

Another clarification on my part: while the discord conversation is an inspiration, I'm not trying to exactly recreate the proposed solution (selecting a previously selected group), but to create a solution (inverting the selection relative to a control group) that solves the same problem (selecting two different halves of the same control group).

**Comment by sprunk** [comment]
Sure, sounds good.
